rir/src/compiler/pir/stack_machine.cpp

```diff
@@ -0,0 +1,397 @@
+            break
+                BINOP(LOr, lgl_or_);
+                BINOP(LAnd, lgl_and_);
+                BINOP(Lt, lt_);
+                BINOP(Gt, gt_);
+                BINOP(Gte, le_);
+                BINOP(Lte, ge_);
+                BINOP(Mod, mod_);
+                BINOP(Div, div_);
+                BINOP(IDiv, idiv_);
+                BINOP(Add, add_);
+                BINOP(Mul, mul_);
+                BINOP(Colon, colon_);
+                BINOP(Pow, pow_);
+                BINOP(Sub, sub_);
+                BINOP(Eq, eq_);
+                BINOP(Neq, ne_);
+
+        #undef BINOP
+        #define UNOP(Name, Op)                                                         \
+        case Opcode::Op:                                                           \
+            v = pop();                                                             \
+            push((*builder)(new Name(v)));                                                  \
+            break
+                UNOP(Plus, uplus_);
+                UNOP(Minus, uminus_);
+                UNOP(Inc, inc_);
+                UNOP(Not, not_);
+                UNOP(Is, is_);
+                UNOP(Length, length_);
+        #undef UNOP
+
+        case Opcode::pick_:
+            push(at(bc.immediate.i));
+            break;
+
+        case Opcode::put_:
+            x = top();
+            for (size_t i = 0; i < bc.immediate.i - 1; ++i)
+                set(i, at(i + 1));
+            set(bc.immediate.i, x);
+            break;
+
+        // TODO implement!
+        // (silently ignored)
+        case Opcode::set_shared_:
+        case Opcode::invisible_:
+        case Opcode::visible_:
+        case Opcode::isfun_:
+        case Opcode::make_unique_:
+        case Opcode::brobj_:
+            break;
+
+        // Currently unused opcodes:
+        case Opcode::ldarg_:
+        case Opcode::alloc_:
+        case Opcode::push_code_:
+        case Opcode::set_names_:
+        case Opcode::names_:
+        case Opcode::force_:
+
+        // Invalid opcodes:
+        case Opcode::label:
+        case Opcode::invalid_:
+        case Opcode::num_of:
+
+        // Opcodes handled elsewhere
+        case Opcode::brtrue_:
+        case Opcode::brfalse_:
+        case Opcode::br_:
+            assert(false);
+
+        // Unsupported opcodes:
+        case Opcode::ldloc_:
+        case Opcode::stloc_:
+        case Opcode::ldlval_:
+        case Opcode::asast_:
+        case Opcode::missing_:
+        case Opcode::dispatch_stack_:
+        case Opcode::dispatch_:
+        case Opcode::guard_env_:
+        case Opcode::call_stack_:
+        case Opcode::return_:
+        case Opcode::beginloop_:
+        case Opcode::endcontext_:
+        case Opcode::ldddvar_:
+        case Opcode::int3_:
+            std::cerr << "Cannot compile Function. Unsupported bc\n";
+            bc.print();
+            assert(false);
+            break;
+    }
+    this->consumeBC(src->body()->endCode());
```

Comment:
ok, found your bug! when you are compiling promises, then this compares the pc from the promise to the end of the function. (ie. in Rir2Pir srcCode vs. srcFunction)